[PATCH v14] API: add event vectors and event aggregators

JannePeltonen opened the pull request:
Introduce event vectors that, similar to packet vectors, are events that contain multiple events. An event vector can contain multiple types of events, not just packet events.

Introduce event aggregators which are associated with an underlying queue, get created when the underlying queue is created and which have their own queue handle that can be used for enqueueing purposes.

This PR replaces PR 2149. Event vectors are taken directly from that PR but put to a different commit from event aggregators. Event aggregators are based on the work in PR 2148, but now the aggregators can be referred to using queue handles gotten from odp_queue_aggregator_get() function. The start-of-vector/end-of-vector flags are not included in this PR at this point.

JannePeltonen commented:
v2:
- added capabilities for max number of aggregators and max number of aggregators per queue
- added possibility to have no requested timeout in which case an implementation may queue packets indefinitely
- renamed odp_queue_aggregator_get() to odp_queue_aggregator()

PavanNikhilesh commented on include/odp/api/spec/queue_types.h:
```diff
@@ -214,6 +226,9 @@ typedef struct odp_queue_capability_t {
 		} waitfree;
 
+		/** Event vector generation capabilities */
+		odp_event_aggregator_capability_t aggregator;
```
I don't think aggregator should be a part of pail queue capability.

MatiasElo commented on include/odp/api/spec/queue.h:
```diff
@@ -155,6 +158,50 @@ int odp_queue_context_set(odp_queue_t queue, void *context, uint32_t len);
+ *
+ * The aggregator queues have the same enq_mode as the underlying queue.
+ *
+ * The returned queue handle cannot be used for dequeuing events. It must
+ * not be passed to odp_queue_deq() or similar.
+ *
+ * The returned queue handle must not be passed to any of the
+ * odp_queue_context* or odp_queue_sched* functions and is never returned
+ * by odp_queue_lookup().
```
Perhaps the forbidden functions could be listed explicitly in the same manner than for example in `odp_init_global()`.

JannePeltonen replied:
I did not yet change this.

JannePeltonen replied:
Clarified in v5.

MatiasElo commented on include/odp/api/abi-default/event_types.h:
```diff
@@ -1,6 +1,6 @@
 /* SPDX-License-Identifier: BSD-3-Clause
  * Copyright (c) 2017-2018 Linaro Limited
- * Copyright (c) 2022-2023 Nokia
+ * Copyright (c) 2022-2024 Nokia
```
Copyright years can be updated.

JannePeltonen replied:
I updated copyrights only in the files that actually got changes this year. For this particular file, 2022-2024 is still valid.

JannePeltonen commented on include/odp/api/spec/event_vector_types.h:
```diff
@@ -33,6 +33,86 @@ extern "C" {
+ * Event aggregator capabilities
+ */
+typedef struct odp_event_aggregator_capability_t {
+	/** Maximum number of event aggregators for this queue type */
+	uint32_t max_num_aggregators;
+
+	/** Maximum number of event aggregators per queue */
+	uint32_t max_num_aggregators_per_queue;
```
fixed in v3

JannePeltonen commented on include/odp/api/spec/queue_types.h:
```diff
@@ -214,6 +226,9 @@ typedef struct odp_queue_capability_t {
 
 		} waitfree;
 
+		/** Event vector generation capabilities */
+		odp_event_aggregator_capability_t aggregator;
```
fixed in v3, the 'vector' name was not intentional.

JannePeltonen commented on include/odp/api/spec/queue.h:
```diff
@@ -35,6 +35,9 @@ extern "C" {
  * parameters into their default values. Default values are also used when
  * 'param' pointer is NULL.
  *
+ * 'name' and 'param' and any configuration structures pointed to by 'param'
+ * are used only during the call. They can be freed after queue creation.
+ *
```
I did not (yet?) change this. This is a bit different from the typical case that there is not only a pointer to the paramer structure but a pointer to an other objet (the array) within the parameter structure. It is not very clear in the API what the lifetime of such objects are so I added this remark here. I would actually prefer keeping it until there is an adequate generic comment in a single place (if there can be). We have the same issue in some other APIs as well (e.g. it is not obvious that crypto keys passed to crypto or IPsec do not need to outlive the session/SA creation call).

JannePeltonen replied:
We decided to handle this in a general way in a separate commit, so removing the remark in v4.

JannePeltonen commented on include/odp/api/spec/queue.h:
```diff
@@ -155,6 +158,50 @@ int odp_queue_context_set(odp_queue_t queue, void *context, uint32_t len);
+ * the event aggregator.
+ *
+ * This function does not create a new queue but merely returns a reference
+ * to an aggregator queue which has the same life time as the underlying
```
fixed in v3

JannePeltonen commented on include/odp/api/spec/queue_types.h:
```diff
@@ -291,6 +306,28 @@ typedef struct odp_queue_param_t {
+	  * When >= 1, configuration must be provided for each aggregator
+	  * through the 'aggregator' array.
+	  */
+	uint32_t num_aggregators;
```
In v3 I added a statement about this in the desctiption of odp_queue_aggregator().

JannePeltonen commented:
v3: addresses most of the review comments so far.

Also added a draft commit that adds crypto and IPsec capabilities that indicate if aggregator queues are supported. The commit is just a draft for discussion at this point since it is not clear to me if such capabilities are desirable and what their semantics should be. I guess the choise is between:
- Aggregator queues can be used in (crypto/ipsec/pktio/DMA/ML/timer/...) only if capa says so
- Aggregator queues can be used always, and capa indicates whether they might have an effect
- Aggregator queues can be used always but may not have an effect (the underlying queue is used instead) and there are no capa bits
- API texts says where aggregator queues can be used.

JannePeltonen commented:
v4:
	- removed the draft commit that introduced crypto and IPsec capabilities for aggregation support
	- clarified that aggregation queue handle can be given to ODP event sources even if aggregation is not supported.
	- mentioned that order is not preserved between events going through different aggregators
	- specified default values for aggregator specific queue parameters
	- removed the remark about queue creation parameter lifetimes

JannePeltonen commented:
v5:
- added a separate pool type for event vectors, keeping the old pool type for packet vectors only
- shortened 'aggregator' to 'aggr' in code
- small documentation fixes

psavol commented on include/odp/api/spec/event_vector.h:
```diff
@@ -0,0 +1,249 @@
+ * Check user flag
+ *
+ * Implementation clears user flag during new event vector creation (e.g. alloc
+ * and event input) and reset. User may set the flag with
```
"event input" should be "packet input"

JannePeltonen replied:
fixed in v6

psavol commented on include/odp/api/spec/event_vector.h:
```diff
@@ -0,0 +1,249 @@
+ * @retval 0 Event vector is not valid
+ * @retval 1 Event vector is valid
+ */
+int odp_event_vector_valid(odp_event_vector_t evv);
```
This may not be needed. Today "is valid" functions are defined only to event, buffer, packet and packet_vector. "to_u64" and debug print functions could be enough for debugging.

JannePeltonen replied:
removed from v6

psavol commented on include/odp/api/spec/event_vector_types.h:
```diff
@@ -33,6 +33,86 @@ extern "C" {
+	uint32_t min_size;
+
+	/** Maximum allowed value of odp_event_aggr_confit_t::max_tmo_ns */
+	uint64_t max_tmo_ns;
```
Typo: confit_t -> config_t

JannePeltonen replied:
fixed in v6

psavol commented:
V5 API spec is good. Found two typos and had a question about "is valid" function.

JannePeltonen commented:
v6:
- Clarified the text that describes where an aggregator queue handle cannot be used
- Removed odp_event_vector_valid()
- Fixed typos

psavol commented on include/odp/api/spec/pool_types.h:
```diff
@@ -613,6 +648,27 @@ typedef struct odp_pool_param_t {
+		uint32_t max_size;
+
+		/** Minimum user area size in bytes. The maximum value is defined by
+		 *  pool capability vector.max_uarea_size. Specify as 0 if no user
```
Typo: vector.max_uarea_size -> event_vector.max_uarea_size

JannePeltonen commented:
v7: fixed a couple of typos

PavanNikhilesh reviewed (approved):
Acked-by: Pavan Nikhilesh <pbhagavatula@marvell.com>

MatiasElo commented on include/odp/api/spec/queue.h:
```diff
@@ -249,7 +251,7 @@ odp_schedule_sync_t odp_queue_sched_type(odp_queue_t queue);
  * Queue priority
  *
- * @note Passing an invalid queue_handle will result in UNDEFINED behavior
+ * @note The queue must be of type ODP_QUEUE_TYPE_SCHED
```
Could use `#ODP_QUEUE_TYPE_SCHED` to enable Doxygen link generation.

JannePeltonen replied:
v9

JannePeltonen commented:
v9: minor doxygen improvement, typo fix and review tags added

JannePeltonen commented:
v11:
Changes in the API commits:
- added missing #include <odp/api/event_vector.h> in odp_api.h
- added missing ODP_EVENT_ANY definition in the abi headers
- updated a few copyright years

New commits added:
- added event vector implementation and tests
- added minimal event aggregator implementation (zero aggregators) and a few tests (excluding tests for the actual aggregation function)
- added printing of new capabilities in sysinfo example app

MatiasElo commented on platform/linux-generic/include-abi/odp/api/abi/packet.h:
```diff
@@ -17,6 +17,7 @@ extern "C" {
 #endif
 
 #include <odp/api/plat/packet_inlines.h>
+#include <odp/api/plat/event_vector_inlines.h>
```
This belongs to `platform/linux-generic/include-abi/odp/api/abi/event_vector.h`

JannePeltonen replied:
moved in v12

MatiasElo commented on test/validation/api/event/event_vector.c:
```diff
@@ -0,0 +1,481 @@
+	static odp_pool_param_t param; /* keep around until tests finished */
+
+	if (odp_pool_capability(&capa) != 0) {
+		printf("\nodp_pool_capability() failed\n");
```
Generally the validation tests use `ODPH_ERR()` macro.

JannePeltonen replied:
changed in v12

MatiasElo commented on test/validation/api/event/event_vector.c:
I'd prefer moving the event vector tests into a different directory (`test/validation/api/event_vector`), as event vector is a separate API module. Test failures can also be identified a bit faster this way.

JannePeltonen replied:
I did not do the move in v12 as I am not sure it really improves things and it will slow down overall build time just a little because of the linking needed for the new test app. The tests are anyway already in their own C file and test suite.

JannePeltonen commented on test/validation/api/queue/queue.c:
```diff
@@ -1152,6 +1156,147 @@ static void queue_test_mt_plain_nonblock_lf(void)
+	queue = odp_queue_create("queue_test_aggr", &param);
+	CU_ASSERT_FATAL(queue != ODP_QUEUE_INVALID);
+	CU_ASSERT(odp_queue_aggr(queue, 0) == ODP_QUEUE_INVALID);
+	odp_queue_destroy(queue);
```
checked in v12

JannePeltonen commented on test/validation/api/queue/queue.c:
```diff
@@ -1152,6 +1156,147 @@ static void queue_test_mt_plain_nonblock_lf(void)
+{
+	odp_queue_capability_t capa;
+
+	odp_queue_capability(&capa);
```
checked in v12

JannePeltonen commented on test/validation/api/queue/queue.c:
```diff
@@ -1152,6 +1156,147 @@ static void queue_test_mt_plain_nonblock_lf(void)
+			  odp_queue_to_u64(queue));
+		if (i == 0 || i == num - 1)
+			odp_queue_print(aggr_queue);
+		CU_ASSERT(odp_queue_info(aggr_queue, &info));
```
fixed in v12

JannePeltonen commented:
v12: updates in the non-API commits based on review comments

MatiasElo reviewed (approved):
OK, assuming the CI passes.

JannePeltonen commented:
v13: added review tags

JannePeltonen commented:
v14: rebased
